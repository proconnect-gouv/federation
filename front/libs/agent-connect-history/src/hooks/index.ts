/* istanbul ignore file */

// declarative file
export * from './use-add-to-user-history.hook';
export * from './use-remove-from-user-history.hook';
export * from './use-user-history.hook';
