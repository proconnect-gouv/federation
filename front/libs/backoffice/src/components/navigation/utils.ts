import { NavigationItem } from '@fc/routing';

export const filterRouteHasOrderProperty = (
  route: NavigationItem,
): route is Required<NavigationItem> => {
  if (route.order === undefined) return false;
  return route.order >= 0;
};

export const sortNavigationRouteByOrder = (
  a: Required<NavigationItem>,
  b: Required<NavigationItem>,
): number => a.order - b.order;
