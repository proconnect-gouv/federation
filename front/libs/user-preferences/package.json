{
  "name": "@fc/user-preferences",
  "version": "0.1.0",
  "private": true,
  "dependencies": {
    "@fc/backoffice": "^0.1.0",
    "classnames": "^2.3.1",
    "final-form": "^4.20.6",
    "react": "^17.0.2",
    "react-final-form": "^6.5.7",
    "react-final-form-listeners": "^1.0.3"
  },
  "devDependencies": {
    "@fc/tests-utils": "^0.1.0",
    "@types/react-final-form-listeners": "^1.0.0"
  },
  "main": "src/index.ts",
  "browserslist": [
    ">0.2%",
    "not dead",
    "not op_mini all"
  ]
}
