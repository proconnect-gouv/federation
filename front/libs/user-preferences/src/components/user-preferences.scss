@import '@fc/dsfr/index.scss';

.ServicesListComponent {
  .with-border {
    border-top: 1px solid #eeeeee;
  }
}

.ServiceComponent {
  &-image {
    background: #e5edf5;
    border-radius: 4px;
    height: 70px;
    max-height: 70px;
    max-width: 232px;
    min-height: 70px;
    min-width: 232px;
    width: 232px;

    img {
      display: block;
      left: 50%;
      max-height: 76%;
      max-width: 84%;
      position: absolute;
      top: 50%;
      transform: translate(-50%, -50%);
    }
  }

  .FieldSwitchInputComponent {
    &-toggle-button {
      background: $error;
    }

    &-toggle-thumb {
      background: $white;
    }

    &-toggle-thumb,
    &-toggle-button {
      box-shadow: inset 0 0 0 1px $error;
    }

    &-icon,
    &-legend {
      color: $error;
      font-weight: bold;
    }
  }

  .FieldSwitchInputComponent.checked {
    .FieldSwitchInputComponent {
      &-toggle-button {
        background: $success;
      }

      &-toggle-thumb,
      &-toggle-button {
        box-shadow: inset 0 0 0 1px $success;
      }

      &-icon,
      &-legend {
        color: $success;
      }
    }
  }
}
