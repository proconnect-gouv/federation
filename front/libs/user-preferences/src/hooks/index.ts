/* istanbul ignore file */

// declarative file
export * from './use-user-preferences-api.hook';
