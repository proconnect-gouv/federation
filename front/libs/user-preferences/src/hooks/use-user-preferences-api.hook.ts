import axios, { AxiosError, AxiosResponse } from 'axios';
import { useCallback, useEffect, useState } from 'react';

import { useApiGet } from '@fc/common';

import {
  FormValues,
  IGetCsrfTokenResponse,
  UserPreferencesConfig,
  UserPreferencesData,
} from '../interfaces';
import { UserPreferencesService, UserPreferencesServiceInterface } from '../services';

export const validateHandlerCallback = ({ idpList }: Pick<FormValues, 'idpList'>) => {
  const isDefinedPreferences = idpList && Object.values(idpList);
  const hasError = isDefinedPreferences && !isDefinedPreferences.includes(true);

  if (!hasError) {
    return undefined;
  }

  return {
    idpList: 'error',
  };
};

export const useUserPreferencesApi = (options: UserPreferencesConfig) => {
  const userPreferences: UserPreferencesData = useApiGet({
    endpoint: options.API_ROUTE_USER_PREFERENCES,
  });

  const [submitErrors, setSubmitErrors] = useState<AxiosError | Error | undefined>(undefined);
  const [submitWithSuccess, setSubmitWithSuccess] = useState(false);
  const [formValues, setFormValues] = useState<FormValues | undefined>(undefined);

  const validateHandler = useCallback(validateHandlerCallback, []);

  const commitErrorHandler = useCallback((err: AxiosError | Error) => {
    setSubmitErrors(err);
    setSubmitWithSuccess(false);
  }, []);

  const commitSuccessHandler = useCallback(({ data }: AxiosResponse<UserPreferencesData>) => {
    const { allowFutureIdp, idpList } = UserPreferencesService.parseFormData(data);
    setFormValues({ allowFutureIdp, idpList });
    setSubmitErrors(undefined);
    setSubmitWithSuccess(true);
  }, []);

  const commit = useCallback(
    async ({
      allowFutureIdp,
      idpList,
    }: Pick<UserPreferencesServiceInterface, 'allowFutureIdp' | 'idpList'>) => {
      const {
        data: { csrfToken },
      } = await axios.get<IGetCsrfTokenResponse>(options.API_ROUTE_CSRF_TOKEN);

      const data = UserPreferencesService.encodeFormData({
        allowFutureIdp,
        csrfToken,
        idpList,
      });

      return axios
        .post<UserPreferencesData>(options.API_ROUTE_USER_PREFERENCES, data)
        .then(commitSuccessHandler)
        .catch(commitErrorHandler);
    },
    [
      commitSuccessHandler,
      commitErrorHandler,
      options.API_ROUTE_CSRF_TOKEN,
      options.API_ROUTE_USER_PREFERENCES,
    ],
  );

  useEffect(() => {
    if (userPreferences && !formValues) {
      // @NOTE
      // executed at first render only
      // when
      // if initial userPreferences has already been loaded
      // and if form values has not yet been set
      const { allowFutureIdp, idpList } = UserPreferencesService.parseFormData(userPreferences);
      setFormValues({ allowFutureIdp, idpList });
    }
  }, [userPreferences, formValues]);

  return {
    commit,
    formValues,
    submitErrors,
    submitWithSuccess,
    userPreferences,
    validateHandler,
  };
};
