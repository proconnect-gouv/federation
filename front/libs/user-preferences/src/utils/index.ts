/* istanbul ignore file */

// declarative file
export * from './user-preferences-form.util';
