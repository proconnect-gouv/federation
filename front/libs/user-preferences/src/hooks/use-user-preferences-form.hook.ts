import { useCallback, useEffect, useState } from 'react';

import { UserPreferencesData } from '../interfaces';
import { checkHasDefaultConfiguration, checkSomeIdpHasBeenChangedSinceLoading } from '../utils';

export interface useUserPreferencesFormProps {
  dirtyFields: Record<string, boolean>;
  userPreferences: UserPreferencesData;
}

export const useUserPreferencesForm = ({
  dirtyFields,
  userPreferences,
}: useUserPreferencesFormProps) => {
  const hasDefaultConfiguration = checkHasDefaultConfiguration(userPreferences);
  const someIdpHasBeenChangedSinceLoading = checkSomeIdpHasBeenChangedSinceLoading(dirtyFields);

  const [alertInfoState, setAlertInfoState] = useState({
    hasInteractedWithAlertInfo: false,
    isDisplayedAlertInfo: false,
  });

  const allowingIdPConfirmation = useCallback(() => {
    // @NOTE can not be mocked without a native re-implementation
    // istanbul ignore next
    setAlertInfoState((previous) => ({
      ...previous,
      hasInteractedWithAlertInfo: true,
      isDisplayedAlertInfo: false,
    }));
  }, [setAlertInfoState]);

  useEffect(() => {
    const shouldDisplayAlert =
      !alertInfoState.hasInteractedWithAlertInfo &&
      hasDefaultConfiguration &&
      someIdpHasBeenChangedSinceLoading;

    if (shouldDisplayAlert) {
      setAlertInfoState((previous) => ({
        ...previous,
        isDisplayedAlertInfo: true,
      }));
    }
  }, [
    someIdpHasBeenChangedSinceLoading,
    alertInfoState.hasInteractedWithAlertInfo,
    hasDefaultConfiguration,
  ]);

  return {
    alertInfoState,
    allowingIdPConfirmation,
  };
};
