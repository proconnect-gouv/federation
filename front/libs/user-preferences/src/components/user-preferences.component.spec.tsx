import { render } from '@testing-library/react';
import { mocked } from 'ts-jest/utils';

import { useUserPreferencesApi } from '../hooks';
import { UserPreferencesComponent } from './user-preferences.component';
import { UserPreferencesFormComponent } from './user-preferences-form.component';

jest.mock('react-final-form');
jest.mock('../hooks');
jest.mock('./services-list.component');
jest.mock('./user-preferences-form.component');

describe('UserPreferencesComponent', () => {
  // given
  const optionsMock = {
    API_ROUTE_CSRF_TOKEN: 'csrf-token-endpoint',
    API_ROUTE_USER_PREFERENCES: 'any-endpoint',
  };
  const commitMock = jest.fn();
  const initialValuesMock = { allowFutureIdp: false, idpList: expect.any(Object) };
  const userPreferencesMock = {
    allowFutureIdp: false,
    idpList: [expect.any(Object), expect.any(Object)],
  };
  const hookResultMock = {
    commit: commitMock,
    formValues: initialValuesMock,
    submitErrors: undefined,
    submitWithSuccess: false,
    userPreferences: userPreferencesMock,
  };

  beforeEach(() => {
    jest.clearAllMocks();
    mocked(useUserPreferencesApi).mockReturnValue(hookResultMock);
  });

  it('should match the snapshot', () => {
    // when
    const { container } = render(<UserPreferencesComponent options={optionsMock} />);
    // then
    expect(container).toMatchSnapshot();
  });

  it('should useUserPreferencesApi have been called', () => {
    // when
    render(<UserPreferencesComponent options={optionsMock} />);
    // then
    expect(useUserPreferencesApi).toHaveBeenCalledTimes(1);
    expect(useUserPreferencesApi).toHaveBeenCalledWith(optionsMock);
  });

  it.skip('should call UserPreferencesFormComponent with default params', () => {
    // when
    render(<UserPreferencesComponent options={optionsMock} />);
    // then
    expect(UserPreferencesFormComponent).toHaveBeenCalledTimes(1);
    expect(UserPreferencesFormComponent).toHaveBeenCalledWith(
      {
        canNotSubmit: expect.any(Boolean),
        onSubmit: expect.any(Function),
        showNotification: false,
        userPreferences: userPreferencesMock,
      },
      {},
    );
  });
});
