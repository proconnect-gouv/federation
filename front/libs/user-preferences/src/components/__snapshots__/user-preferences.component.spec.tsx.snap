// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserPreferencesComponent should match the snapshot 1`] = `
<div>
  <div>
    UserPreferencesFormComponent
  </div>
</div>
`;
